File: apps/api/src/facilio/schemas/transformations.py
"""Version and transformation API schemas."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field

from facilio.schemas.datasets import DatasetSummary

VersionKind = Literal["ORIGINAL", "DERIVED"]
ProfileStatus = Literal["NOT_PROFILED", "PROFILING", "READY", "FAILED"]


class DatasetVersionSummary(BaseModel):
    id: UUID
    dataset_id: UUID
    version_number: int
    parent_version_id: UUID | None = None
    kind: VersionKind
    label: str
    row_count: int | None = None
    column_count: int | None = None
    profile_status: ProfileStatus
    profiled_at: datetime | None = None
    created_at: datetime
    is_current: bool = False
    operation_code: str | None = None
    operation_summary: str | None = None
    created_by_workflow_run_id: UUID | None = None
    workflow_name: str | None = None
    workflow_revision: int | None = None


class DatasetVersionDetail(DatasetVersionSummary):
    columns: list[dict[str, Any]] = Field(default_factory=list)


class TransformationRequest(BaseModel):
    operation: str = Field(min_length=1, max_length=64)
    parameters: dict[str, Any] = Field(default_factory=dict)


class SetCurrentVersionRequest(BaseModel):
    version_id: UUID


class TransformationImpactData(BaseModel):
    rows_before: int
    rows_after: int
    columns_before: int
    columns_after: int
    changed_cell_count: int
    removed_row_count: int
    removed_column_count: int
    affected_row_count: int
    no_op: bool


class ChangeExampleData(BaseModel):
    kind: Literal["cell", "row_removed", "column_removed"]
    row_index: int | None = None
    column: str | None = None
    before: Any = None
    after: Any = None
    reason: str | None = None


class TransformationPreviewData(BaseModel):
    operation: str
    parameters: dict[str, Any]
    input_version_id: UUID
    impact: TransformationImpactData
    examples: list[ChangeExampleData] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    summary: str
    extra: dict[str, Any] = Field(default_factory=dict)


class QualityDeltaDimension(BaseModel):
    key: str
    label: str
    before: float | None = None
    after: float | None = None
    delta: float | None = None


class QualityDeltaData(BaseModel):
    before: float | None = None
    after: float | None = None
    delta: float | None = None
    dimensions: list[QualityDeltaDimension] = Field(default_factory=list)


class TransformationApplyData(BaseModel):
    version: DatasetVersionDetail
    transformation_id: UUID
    summary: str
    impact: TransformationImpactData
    profile_status: ProfileStatus
    quality_delta: QualityDeltaData | None = None


class TransformationRecordData(BaseModel):
    id: UUID
    dataset_id: UUID
    input_version_id: UUID
    output_version_id: UUID
    operation_code: str
    parameters: dict[str, Any]
    summary: str
    impact: TransformationImpactData
    created_at: datetime


class LineageNodeData(BaseModel):
    version: DatasetVersionSummary
    transformation: TransformationRecordData | None = None


class LineageData(BaseModel):
    items: list[LineageNodeData]


class VersionComparisonData(BaseModel):
    parent: DatasetVersionSummary
    child: DatasetVersionSummary
    impact: TransformationImpactData | None = None
    transformation: TransformationRecordData | None = None
    quality_delta: QualityDeltaData | None = None


class WorkspaceStatsData(BaseModel):
    datasets: int
    derived_versions: int
    transformations_applied: int
    datasets_analyzed: int
    workflow_count: int = 0
    workflow_run_count: int = 0
    successful_run_count: int = 0
    failed_run_count: int = 0
    queued_job_count: int = 0
    running_job_count: int = 0
    failed_job_count: int = 0
    user_dataset_count: int = 0
    sample_dataset_count: int = 0
    recent_datasets: list[DatasetSummary] = Field(default_factory=list)

File: apps/web/src/types/transformations.ts
export type VersionKind = "ORIGINAL" | "DERIVED";

export type TransformationCategory = "CLEAN_TEXT" | "MISSING_DATA" | "ROWS" | "COLUMNS";

export interface ParameterSpec {
  name: string;
  type: string;
  required: boolean;
  description: string;
  options: string[] | null;
  default: unknown;
}

export interface TransformationDefinition {
  code: string;
  display_name: string;
  description: string;
  category: TransformationCategory;
  supported_column_types: string[];
  dataset_level: boolean;
  notes: string[];
  parameters: ParameterSpec[];
}

export interface DatasetVersion {
  id: string;
  dataset_id: string;
  version_number: number;
  parent_version_id: string | null;
  kind: VersionKind;
  label: string;
  row_count: number | null;
  column_count: number | null;
  profile_status: "NOT_PROFILED" | "PROFILING" | "READY" | "FAILED";
  profiled_at: string | null;
  created_at: string;
  is_current: boolean;
  operation_code: string | null;
  operation_summary: string | null;
  created_by_workflow_run_id: string | null;
  workflow_name: string | null;
  workflow_revision: number | null;
}

export interface DatasetVersionDetail extends DatasetVersion {
  columns: { name: string; index: number; dtype: string }[];
}

export interface TransformationImpact {
  rows_before: number;
  rows_after: number;
  columns_before: number;
  columns_after: number;
  changed_cell_count: number;
  removed_row_count: number;
  removed_column_count: number;
  affected_row_count: number;
  no_op: boolean;
}

export interface ChangeExample {
  kind: "cell" | "row_removed" | "column_removed";
  row_index: number | null;
  column: string | null;
  before: unknown;
  after: unknown;
  reason: string | null;
}

export interface TransformationPreview {
  operation: string;
  parameters: Record<string, unknown>;
  input_version_id: string;
  impact: TransformationImpact;
  examples: ChangeExample[];
  warnings: string[];
  summary: string;
  extra: Record<string, unknown>;
}

export interface QualityDeltaDimension {
  key: string;
  label: string;
  before: number | null;
  after: number | null;
  delta: number | null;
}

export interface QualityDelta {
  before: number | null;
  after: number | null;
  delta: number | null;
  dimensions: QualityDeltaDimension[];
}

export interface TransformationRecord {
  id: string;
  dataset_id: string;
  input_version_id: string;
  output_version_id: string;
  operation_code: string;
  parameters: Record<string, unknown>;
  summary: string;
  impact: TransformationImpact;
  created_at: string;
}

export interface TransformationApplyResult {
  version: DatasetVersionDetail;
  transformation_id: string;
  summary: string;
  impact: TransformationImpact;
  profile_status: DatasetVersion["profile_status"];
  quality_delta: QualityDelta | null;
}

export interface LineageNode {
  version: DatasetVersion;
  transformation: TransformationRecord | null;
}

export interface Lineage {
  items: LineageNode[];
}

export interface VersionComparison {
  parent: DatasetVersion;
  child: DatasetVersion;
  impact: TransformationImpact | null;
  transformation: TransformationRecord | null;
  quality_delta: QualityDelta | null;
}

export interface WorkspaceStats {
  datasets: number;
  derived_versions: number;
  transformations_applied: number;
  datasets_analyzed: number;
  workflow_count: number;
  workflow_run_count: number;
  successful_run_count: number;
  failed_run_count: number;
  queued_job_count: number;
  running_job_count: number;
  failed_job_count: number;
  user_dataset_count?: number;
  sample_dataset_count?: number;
  recent_datasets?: import("@/types/dataset").DatasetSummary[];
}

export interface SuggestedOperation {
  code: string;
  display_name: string;
  reason: string;
  parameters: Record<string, unknown>;
}

export interface TransformationRequest {
  operation: string;
  parameters: Record<string, unknown>;
}
